scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java

```diff
@@ -3,151 +3,205 @@
 import android.app.Activity;
 import android.graphics.drawable.ColorDrawable;
 import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
 import android.util.TypedValue;
 import android.view.Gravity;
-import android.view.MotionEvent;
+import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
-import android.widget.BaseAdapter;
-import android.widget.ListView;
 import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.DividerItemDecoration;
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+
+import io.flutter.embedding.android.FlutterView;
+import io.flutter.embedding.engine.FlutterEngine;
+import io.flutter.embedding.engine.FlutterEngineCache;
+import io.flutter.embedding.engine.dart.DartExecutor;
 import io.flutter.plugin.common.EventChannel;
-import io.flutter.view.FlutterMain;
-import io.flutter.view.FlutterRunArguments;
-import io.flutter.view.FlutterView;
 
 public class ScrollOverlayActivity extends Activity {
+    private static final String VELOCITY_CHANNEL = "scroll_overlay.flutter.io/velocity";
+    private static final int[] OVERLAY_COLORS = new int[]{0x40ff0000, 0x4000ff00, 0x400000ff};
+    private static final String engineId = "engine_id";
 
-    private static final String TAG = "ScrollOverlayActivity";
+    FlutterEngine flutterEngine;
 
-    private static final String VELOCITY_CHANNEL = "scroll_overlay.flutter.io/velocity";
+    class VelocityTracker extends RecyclerView.OnScrollListener {
+        int currentDy = 0;
+        boolean locked = false;
+        EventChannel.EventSink velocitySink;
 
-    private static final int[] OVERLAY_COLORS = new int[]{0x40ff0000, 0x4000ff00, 0x400000ff};
+        /**
+         * How many times the velocity is measured per second.
+         * <p>
+         * Setting this to not too small value - to get a meaningful velocity information,
+         * and not too big - to distinguish individual digits after thousands.
+         */
+        static final int measurementsPerSecond = 25;
 
-    private FlutterView flutterView;
+        final Handler handler = new Handler(Looper.getMainLooper());
+        Runnable velocityTrackerRunnable = new Runnable() {
+            @Override
+            public void run() {
+                if (flutterEngine == null) {
+                    return;
+                }
+                if (!locked) {
+                    velocitySink.success((double) currentDy * measurementsPerSecond);
+                    if (currentDy == 0) {
+                        locked = true;
+                    }
+                    currentDy = 0;
+                }
+                handler.postDelayed(velocityTrackerRunnable, 1000 / measurementsPerSecond);
```

Comment:
shouldn't this run on each frame?  e.g. Is it better to schedule this using https://developer.android.com/reference/android/view/Choreographer.FrameCallback ?

Reply:
We don't want it to update too often. The comment I left on `measurementsPerSecond` seems relevant here

```
* Setting this to not too small value - to get a meaningful velocity information,
* and not too big - to distinguish individual digits after thousands.
```